plenum/test/common/test_digest_validation.py

```diff
@@ -44,46 +48,74 @@ def test_send_same_txn_with_different_signatures_in_separate_batches(
-    wh, did = sdk_wallet_stewards[0]
+        looper, txnPoolNodeSet, sdk_pool_handle, two_requests, tconf):
+    req1, req2 = two_requests
 
-    req = json.dumps(sdk_gen_request(op, protocol_version=CURRENT_PROTOCOL_VERSION,
-                                     identifier=did).as_dict)
-    req1 = sdk_multisign_request_object(looper, sdk_wallet_stewards[0], req)
-    req_obj1 = Request(**json.loads(req1))
-
-    req2 = sdk_multisign_request_object(looper, sdk_wallet_stewards[1], req1)
-    req_obj2 = Request(**json.loads(req2))
+    lo_before = txnPoolNodeSet[0].master_replica.last_ordered_3pc[1]
 
-    assert req_obj1.payload_digest == req_obj2.payload_digest
-    assert req_obj1.digest != req_obj2.digest
+    old_reqs = len(txnPoolNodeSet[0].requests)
+    with max_3pc_batch_limits(tconf, size=2):
+        sdk_send_signed_requests(sdk_pool_handle, [req1])
+        sdk_send_signed_requests(sdk_pool_handle, [req2])
 
-    lo_before = txnPoolNodeSet[0].master_replica.last_ordered_3pc[1]
-    sdk_send_signed_requests(sdk_pool_handle, [req1])
-    sdk_send_signed_requests(sdk_pool_handle, [req2])
+        # We need to check for ordering this way, cause sdk do not allow
+        # track two requests with same reqId at the same time
+        def wait_one_batch(before):
+            assert txnPoolNodeSet[0].master_replica.last_ordered_3pc[1] == before + 1
 
-    # We need to check for ordering this way, cause sdk do not allow
-    # track two requests with same reqId at the same time
-    def wait_one_batch(before):
-        assert txnPoolNodeSet[0].master_replica.last_ordered_3pc[1] == before + 1
+        looper.run(eventually(wait_one_batch, lo_before))
 
-    looper.run(eventually(wait_one_batch, lo_before))
+    assert len(txnPoolNodeSet[0].requests) == old_reqs + 2
 
     pps = txnPoolNodeSet[0].master_replica.sentPrePrepares
     pp = pps[pps.keys()[-1]]
     idrs = pp.reqIdr
     assert len(idrs) == 1
-    assert req_obj1.digest in idrs
+    assert Request(**json.loads(req1)).digest in idrs
+
+
+def test_suspicious_primary_send_same_request_with_different_signatures(
+        looper, txnPoolNodeSet, sdk_pool_handle, two_requests):
+    def malicious_dynamic_validation(self, request: Request):
+        self.execute_hook(NodeHooks.PRE_DYNAMIC_VALIDATION, request=request)
+
+        operation = request.operation
+        req_handler = self.get_req_handler(txn_type=operation[TXN_TYPE])
+        req_handler.validate(request)
+
+        self.execute_hook(NodeHooks.POST_DYNAMIC_VALIDATION, request=request)
+
+    assert txnPoolNodeSet[0].master_replica.isPrimary
+    txnPoolNodeSet[0].doDynamicValidation = types.MethodType(malicious_dynamic_validation, txnPoolNodeSet[0])
+
+    req1, req2 = two_requests
+
+    old_view = txnPoolNodeSet[0].viewNo
+    sdk_send_signed_requests(sdk_pool_handle, [req1])
+    sdk_send_signed_requests(sdk_pool_handle, [req2])
+
+    waitForViewChange(looper, txnPoolNodeSet, expectedViewNo=old_view + 1)
```

Comment:
It would be great also to check that the second request wasn't ordered, and that the view change happened because of `PPR_WITH_ORDERED_REQUEST`

Reply:
ok